Python/errors.c

```diff
@@ -944,90 +944,132 @@ PyErr_NewExceptionWithDoc(const char *name, const char *doc,
+    PyObject *moduleName = _PyObject_GetAttrId(exc_type, &PyId___module__);
     if (moduleName == NULL || !PyUnicode_Check(moduleName)) {
+        Py_XDECREF(moduleName);
         PyErr_Clear();
-        if (PyFile_WriteString("<unknown>", f) < 0)
-            goto done;
+        if (PyFile_WriteString("<unknown>", file) < 0) {
+            return NULL;
+        }
     }
     else {
         if (!_PyUnicode_EqualToASCIIId(moduleName, &PyId_builtins)) {
-            if (PyFile_WriteObject(moduleName, f, Py_PRINT_RAW) < 0)
-                goto done;
-            if (PyFile_WriteString(".", f) < 0)
-                goto done;
+            if (PyFile_WriteObject(moduleName, file, Py_PRINT_RAW) < 0) {
+                Py_DECREF(moduleName);
+                return NULL;
+            }
+            Py_DECREF(moduleName);
+            if (PyFile_WriteString(".", file) < 0) {
+                return NULL;
+            }
+        }
+        else {
+            Py_DECREF(moduleName);
         }
     }
     if (className == NULL) {
-        if (PyFile_WriteString("<unknown>", f) < 0)
-            goto done;
+        if (PyFile_WriteString("<unknown>", file) < 0) {
+            return NULL;
+        }
     }
     else {
-        if (PyFile_WriteString(className, f) < 0)
-            goto done;
+        if (PyFile_WriteString(className, file) < 0) {
+            return NULL;
+        }
     }
 
-    if (v && v != Py_None) {
-        if (PyFile_WriteString(": ", f) < 0)
-            goto done;
-        if (PyFile_WriteObject(v, f, Py_PRINT_RAW) < 0) {
+    if (exc_value && exc_value != Py_None) {
+        if (PyFile_WriteString(": ", file) < 0) {
+            return NULL;
+        }
+        if (PyFile_WriteObject(exc_value, file, Py_PRINT_RAW) < 0) {
             PyErr_Clear();
-            if (PyFile_WriteString("<exception str() failed>", f) < 0) {
-                goto done;
+            if (PyFile_WriteString("<exception str() failed>", file) < 0) {
+                return NULL;
             }
         }
     }
-    if (PyFile_WriteString("\n", f) < 0)
+    if (PyFile_WriteString("\n", file) < 0) {
+        return NULL;
+    }
+    Py_RETURN_NONE;
+}
+
+
+/* Call when an exception has occurred but there is no way for Python
+   to handle it.  Examples: exception in __del__ or during GC.
+
+   Call sys.unraisablehook(exc_type, exc_value, exc_tb, obj). */
+void
+PyErr_WriteUnraisable(PyObject *obj)
+{
+    PyObject *exc_type, *exc_value, *exc_tb;
+
+    PyErr_Fetch(&exc_type, &exc_value, &exc_tb);
+
+    _Py_IDENTIFIER(unraisablehook);
+    PyObject *hook = _PySys_GetObjectId(&PyId_unraisablehook);
+    PyObject *res;
+    if (hook != NULL && hook != Py_None) {
+        PyObject *args[4];
+        args[0] = (exc_type ? exc_type : Py_None);
+        args[1] = (exc_value ? exc_value : Py_None);
+        args[2] = (exc_tb ? exc_tb : Py_None);
+        args[3] = (obj ? obj : Py_None);
+        res = _PyObject_FastCall(hook, args, Py_ARRAY_LENGTH(args));
```

Comment:
Would not be better to call `_PyErr_WriteUnraisable()` if the hook set an exception instead of silently ignore it?

Reply:
There is no recursion. I propose to call `_PyErr_WriteUnraisable()`, not `PyErr_WriteUnraisable()`. And pass `hook` as the obj argument.

Reply:
PyErr_WriteUnraisable() now logs exception in custom unraisablehook using _PyErr_WriteUnraisable().